executor/importer/table_import.go

```diff
@@ -497,3 +508,89 @@ func (ti *TableImporter) setLastInsertID(id uint64) {
 		ti.lastInsertID = id
 	}
 }
+
+// CheckDiskQuota checks disk quota.
+func (ti *TableImporter) CheckDiskQuota(ctx context.Context) {
+	var locker sync.Locker
+	lockDiskQuota := func() {
+		if locker == nil {
+			ti.diskQuotaLock.Lock()
+			locker = ti.diskQuotaLock
+		}
+	}
+	unlockDiskQuota := func() {
+		if locker != nil {
+			locker.Unlock()
+			locker = nil
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(CheckDiskQuotaInterval):
+		}
+
+		largeEngines, inProgressLargeEngines, totalDiskSize, totalMemSize := local.CheckDiskQuota(ti.backend, ti.diskQuota)
+		if len(largeEngines) == 0 && inProgressLargeEngines == 0 {
+			continue
+		}
+
+		ti.logger.Warn("disk quota exceeded",
+			zap.Int64("diskSize", totalDiskSize),
+			zap.Int64("memSize", totalMemSize),
+			zap.Int64("quota", ti.diskQuota),
+			zap.Int("largeEnginesCount", len(largeEngines)),
+			zap.Int("inProgressLargeEnginesCount", inProgressLargeEngines))
+
+		lockDiskQuota()
+
+		if len(largeEngines) == 0 {
+			ti.logger.Warn("all large engines are already importing, keep blocking all writes")
+			continue
+		}
+
+		if err := ti.backend.FlushAllEngines(ctx); err != nil {
+			ti.logger.Error("flush engine for disk quota failed, check again later", log.ShortError(err))
+			unlockDiskQuota()
+			continue
+		}
+
+		// at this point, all engines are synchronized on disk.
+		// we then import every large engines one by one and complete.
+		// if any engine failed to import, we just try again next time, since the data are still intact.
+		var importErr error
+		for _, engine := range largeEngines {
+			// Use a larger split region size to avoid split the same region by many times.
+			if err := ti.backend.UnsafeImportAndReset(
+				ctx,
+				engine,
+				ti.regionSplitSize*int64(config.MaxSplitRegionSizeRatio),
+				ti.regionSplitKeys*int64(config.MaxSplitRegionSizeRatio),
+			); err != nil {
+				importErr = multierr.Append(importErr, err)
+			}
+		}
+		if importErr != nil {
```

Comment:
import again might not work too, maybe fail the whole task? can leave a todo

Reply:
The original logic is to retry, and I added a discussion comment.